streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java

```diff
@@ -167,6 +170,7 @@ public StateStore getStateStore(final String name) {
                 forward(child, key, value);
             }
         } finally {
+            recordContext.timestamp = currentTimestamp;
```

Comment:
Should we consider saving/restoring the whole record context, just in case downstream processors have changed other parts of the context, not just the timestamp?

Reply:
Headers work differently. Users would mutate the input record header so they understand that the header changes.

We can still address it (I agree that the current header API is not great) but I would exclude it from this PR. Feel free to create a ticket for it.

Reply:
Ah, this is a good point. I was thinking that downstream processors were directly manipulating the record context, but if they do that, it's clearly their responsibility to restore it when they're done.

I think I won't bother with a ticket right now, it seems like the mutable-headers thing is actually just part of the key and value also being mutable. Plus, there's a bunch of other stuff that's not well defined about header handling in Streams. It seems like altogether too much to scope in a ticket without investing some serious design work up front.

I think the current change is fine as-is. Thanks!